File: day-12/rain_risk.py
input_file = 'day-12/input.txt'

#Complex numbers trick from https://www.reddit.com/r/adventofcode/comments/kbj5me/2020_day_12_solutions/gfhsuai/

directions = dict(zip('NESW', [1j, 1, -1j, -1]))


def get_heading(curr, change):
  direction_string = 'NESW'
  return direction_string[(direction_string.index(curr) + change) % len(direction_string)]

def part1(input):
  deltas = dict(zip('NESW', [(0, 1), (1, 0), (0, -1), (-1, 0)]))

  x = y = 0
  heading = 'E'
  for path in input.splitlines():
    # print(x,y,heading, path)
    direction = path[:1]
    length = int(path [1:])
    if direction == 'F':
      x += (deltas[heading][0] * length)
      y += (deltas[heading][1] * length)
    elif direction == 'L':
      range = length // 90
      heading = get_heading(heading, -range)
    elif direction == 'R':
      range = length // 90
      heading = get_heading(heading, range)
    else:
      x += (deltas[direction][0] * length)
      y += (deltas[direction][1] * length)


  return abs(x)+abs(y)

def part2(input):
  deltas = dict(zip('NESW', [(0, 1), (1, 0), (0, -1), (-1, 0)]))

  waypoint_x = 10
  waypoint_y = 1

  x = y = 0
  for path in input.splitlines():
    # print(x,y,waypoint_x, waypoint_y, path)
    direction = path[:1]
    length = int(path [1:])
    if direction == 'F':
      x += (waypoint_x * length)
      y += (waypoint_y * length)
    elif direction == 'L':
      range = length // 90
      while range > 0:
        waypoint_x, waypoint_y = -waypoint_y, waypoint_x
        range -= 1
    elif direction == 'R':
      range = length // 90
      while range > 0:
        waypoint_x, waypoint_y = waypoint_y, -waypoint_x
        range -= 1
    else:
      waypoint_x += (deltas[direction][0] * length)
      waypoint_y += (deltas[direction][1] * length)


  return abs(x)+abs(y)

  return 0

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-12/rain_risk.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strings"
)

const input_file string = "./day-12/input.txt"

const (
	NORTH   = 'N'
	EAST    = 'E'
	SOUTH   = 'S'
	WEST    = 'W'
	FORWARD = 'F'
	LEFT    = 'L'
	RIGHT   = 'R'
)

type action struct {
	direction rune
	length    int
}

type point struct {
	x, y int
}

func (p point) Add(d point) point {
	return point{p.x + d.x, p.y + d.y}
}

func (p point) Scale(scale int) point {
	return point{p.x * scale, p.y * scale}
}

func (p point) Rotate(r point) point {
	return point{p.y * r.y, p.x * r.x}
}

func (p point) GetManhatten() int {
	return int(math.Abs(float64(p.x))) + int(math.Abs(float64(p.y)))
}

func ReadInput(input string) []action {
	actions := []action{}
	for _, line := range strings.Split(strings.TrimSpace(input), "\n") {
		curr_action := action{}
		fmt.Sscanf(line, "%c%d", &curr_action.direction, &curr_action.length)
		actions = append(actions, curr_action)
	}
	return actions
}

var directions = map[rune]point{
	NORTH: {0, 1},
	EAST:  {1, 0},
	SOUTH: {0, -1},
	WEST:  {-1, 0}}

var rotations = map[rune]point{
	LEFT:  {1, -1},
	RIGHT: {-1, 1}}

func Part1(input string) int {
	ship := point{0, 0}
	heading := directions[EAST]

	for _, action := range ReadInput(input) {
		switch action.direction {
		case FORWARD:
			ship = ship.Add(heading.Scale(action.length))
		case LEFT, RIGHT:
			for r := action.length; r > 0; r -= 90 {
				heading = heading.Rotate(rotations[action.direction])
			}
		case NORTH, EAST, SOUTH, WEST:
			course_change := directions[action.direction].Scale(action.length)
			ship = ship.Add(course_change)
		}
	}
	return ship.GetManhatten()
}

func Part2(input string) int {
	ship := point{0, 0}
	waypoint := point{10, 1}

	for _, action := range ReadInput(input) {
		switch action.direction {
		case FORWARD:
			ship = ship.Add(waypoint.Scale(action.length))
		case LEFT, RIGHT:
			for r := action.length; r > 0; r -= 90 {
				waypoint = waypoint.Rotate(rotations[action.direction])
			}
		case NORTH, EAST, SOUTH, WEST:
			course_change := directions[action.direction].Scale(action.length)
			waypoint = waypoint.Add(course_change)
		}
	}
	return ship.GetManhatten()
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
